### PR TITLE
[WIP] #378 Emoji autocomplete

### PR DESCRIPTION
**Pull request by StaroKep**
Resolving #378 Emoji autocomplete.

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "version": "0.1.0",
   "scripts": {
     "build": "webpack --mode production && node postbuild.js",
-    "start": "webpack-dev-server --mode development",
+    "start": "webpack-dev-server --watch-poll --mode development",
```
Comment (StaroKep) [inline frontend/package.json]: It's important for Ubuntu

**Comment by StaroKep** [comment]
@Reeywhaar hi! Please review this PR.

**Comment by StaroKep** [comment]
@umputun hi! Please, try this implementation. I would be grateful for the comments.

**Comment by umputun** [comment]
Hi. Thx for the PR, however, this is not exactly what I had in mind:

1. No need to have Emoji label/dropdown, all we want it to show completion list at the time someone starting to type `:blah` 
1. Position for the autocompletion unexpected and unusual. It should open the selection right in the place, not outside of the edit area
1. Table-like style is too bold and unnecessary
1. Lack of emoji preview

Pls take a look at how it works here (github) or on giltab. This is the goal and all we want is this:

mee0jvgt4w-20191020-151446

or this:

05q7290k0x-20191020-152046

**Comment by StaroKep** [comment]
Oh, I understand. Thank you. :smile:

**Comment by StaroKep** [comment]
@umputun, hi! Please, try this implementation. I would be grateful for the comments.

**Comment by umputun** [comment]
looking better. Btw, it makes sense to attach screenshots, this will save time and efforts

q8suu-201911-04114459-talhr

My suggestions about style:

- get rid of the fat the outer border, make it slimmer and less visible
- make inside separator just slightly visible 
- change selection backgoround 

smth like this:

hu45c-201911-04115016-79ypv

**Comment by umputun** [comment]
another thing - this PR has conflicts. Before @Reeywhaar can review it you should resolve

**Comment by StaroKep** [comment]
@umputun ok, thx. Maybe tomorrow fix your remarks.

**Comment by umputun** [comment]
ping?

**Comment by StaroKep** [comment]
@umputun sorry. Was busy. I remember this task.

**Comment by StaroKep** [comment]
That so awesome (sarcasm)! The packages were updated! Now, I don't know why my implementation partially breaks! :cry:

**Comment by Reeywhaar** [comment]
@StaroKep what exactly is broken? Yep, between your PR I've updated preact to PreactX. On a minor side react-redux's connect is broken (in terms of types), and I'm bind state manually by creating wrapper component and `useSelector`. On a bright side hooks are available!.

**Comment by StaroKep** [comment]
I have method `setSelectableItemsFilter` in `app/components/dropdown/dropdown.tsx`. It must update state when we use it from `app/components/input/input.tsx`. But it's not happens. Actually, some times `selectableItemsFilter` is undefined in `filterSelectableItems`. But second ago we said to set necessary value.
But when we print the second `:` symbol of emoji the state will update. WTF!

**Comment by umputun** [comment]
@StaroKep @Reeywhaar - any clever ideas what's going on and how to make it working again?

**Comment by StaroKep** [comment]
@umputun @Reeywhaar 
Hello! Today I found time to fix errors. Please, try that implementation.
If you will not have important remarks, let's merge that PR into master.

File: frontend/app/components/dropdown/__item/dropdown__item.tsx [hunk 2]
```diff
@@ -4,10 +4,25 @@ import b from 'bem-react-helper';
...
 export default function DropdownItem(props: RenderableProps<Props> & JSX.HTMLAttributes & { separator?: boolean }) {
-  const { children, separator = false } = props;
+  const { children, separator = false, selectable = false, active = false, onMouseOver, index, onClick } = props;
 
-  return <div className={b('dropdown__item', {}, { separator })}>{children}</div>;
+  const classNames = b('dropdown__item', {}, { separator, selectable, active });
+
+  if (onClick) {
+    return (
+      <button data-id={index} onFocus={onMouseOver} onMouseOver={onMouseOver} className={classNames} onClick={onClick}>
```
Comment (Mavrin) [inline frontend/app/components/dropdown/__item/dropdown__item.tsx]: You can pass as argument selected item instead use DOM attribute.

File: frontend/app/components/dropdown/dropdown.tsx [hunk 3]
```diff
@@ -1,10 +1,13 @@
 /** @jsx createElement */
 import { createElement, Component, createRef } from 'preact';
 import b from 'bem-react-helper';
+import nodeEmoji from 'node-emoji';
```
Comment (Mavrin) [inline frontend/app/components/dropdown/dropdown.tsx]: It looks strange, that common component knows about emojis. Could you invert this dependency?

File: frontend/app/components/dropdown/dropdown.tsx [hunk 4]
```diff
@@ -14,30 +17,174 @@ interface Props {
   onTitleClick?: () => void;
   mix?: string;
   theme: Theme;
-  onOpen?: (root: HTMLDivElement) => unknown;
-  onClose?: (root: HTMLDivElement) => unknown;
+  onOpen?: (root: HTMLDivElement) => {};
+  onClose?: (root: HTMLDivElement) => {};
+  activeSelectableItemID?: number;
+  onDropdownItemClick?: (e: Event) => void;
+  withSelectableItems?: boolean;
```
Comment (Mavrin) [inline frontend/app/components/dropdown/dropdown.tsx]: It can be just items

File: frontend/app/components/dropdown/dropdown.tsx [hunk 5]
```diff
@@ -176,9 +360,31 @@ export default class Dropdown extends Component<Props, State> {
...
+          isHover: true,
+        });
+      }
+    }
+  }
+
   render() {
-    const { title, titleClass, heading, children, mix, theme } = this.props;
-    const { isActive } = this.state;
+    let { children } = this.props;
```
Comment (Mavrin) [inline frontend/app/components/dropdown/dropdown.tsx]: It is not good that you override children depend on props. Now component behavior is not predictable.
For example
```jsx
<Dropdown withSelectableItems=[{title: '2'}]>
  <div>item1</div>
  <div>item2</div>
</Dropdown>
```

File: frontend/app/components/dropdown/dropdown.tsx [hunk 6]
```diff
@@ -195,13 +401,16 @@ export default class Dropdown extends Component<Props, State> {
         </Button>
 
         <div
-          className="dropdown__content"
+          className={b('dropdown__content', undefined, { emoji: isEmojiDropdown })}
```
Comment (Mavrin) [inline frontend/app/components/dropdown/dropdown.tsx]: `undefined` ?

File: frontend/app/components/input/input.tsx [hunk 7]
```diff
@@ -114,13 +135,246 @@ export class Input extends Component<Props, State> {
...
+    const inputText = text || this.state.text;
+    const [start] = this.textAreaRef.current.getSelection(); // Start of selection (cursor position)
+
+    const textStart = inputText.substr(0, start); // Text before cursor position
+
+    // Find position of last space and colon
+    const [lastColonPosition, lastSpacePosition] = this.getLastColonAndSpacePosition(textStart);
+
+    // Go out, if no colon
+    if (!~lastColonPosition) return;
```
Comment (Mavrin) [inline frontend/app/components/input/input.tsx]: `lastColonPosition ===  -1`

**Comment by umputun** [comment]
pinging @StaroKep

**Comment by StaroKep** [comment]
@umputun omg, sorry! I will try to fix remarks this weekend. If I do that, can we reopen PR?

**Comment by umputun** [comment]
np. We have this functionality merged in already (#546)

**Comment by StaroKep** [comment]
@umputun oh, ok. I was slow like a snail.

**Pull request closed without merging**